### PR TITLE
Pcell interfacing update

### PR DESCRIPTION
**Pull request by mkosunen**
This pull request contains fixes that enable correct use of pcells in hierarchical bag layout.

@sporrasm , please elaborate the changes made/problems fixed  if you wish.

### CODE REVIEW

File: run_scripts/start_bag.il [hunk 1]
```diff
@@ -1248,15 +1253,6 @@ procedure( create_layout_helper( cv tech_file inst_list rect_list via_list pin_l
                     if( inst->params != nil then
                         ; create pcell instance
                         obj = dbCreateParamInst(cv inst_cv inst->name inst->loc inst->orient 1 inst->params)
-                        ; execute parameter callbacks
```
Comment (sporrasm) [inline run_scripts/start_bag.il:1251]: For reasons yet unknown to me, calling the callbacks at right after instantiating the layout does NOT work. Basically, the layout does not get updated or gets only partially updated, if the callbacks are ran at this point.

**Comment by mkosunen** [comment]
@ayan-biswas Just to check, is master your current development branch or should I select some other target?

File: run_scripts/start_bag.il [hunk 2]
```diff
@@ -1460,7 +1447,15 @@ procedure( create_layout( lib_name view_name via_tech layout_f "ttt" )
             printf("Creating %s__%s (%s)\n" lib_name cell_name view_name)
             create_layout_helper(cv tech_file inst_list rect_list via_list pin_list path_list
                                  blockage_list boundary_list polygon_list)
-
+
```
Comment (sporrasm) [inline run_scripts/start_bag.il]: Removed running callbacks in create_layout_helper. Instead, after the layout has been instantiated, check whether the instance is a Pcell (1st if-else) and check whether the callback function is defined (2nd if-else). If both are true, run the callbacks. If the callback function is NOT defined, print a warning about possibly broken Pcell layouts. This seems to fix pcell layout generation (e.g. layouts are updated correctly).

File: run_scripts/start_bag.il [hunk 3]
```diff
@@ -1185,15 +1184,21 @@ procedure( create_path_seg_helper(cv lay p0 p1 width start_s end_s)
...
         )
-        when( end_s == "round"
-            eext = width / 2
+        if( end_s == "round" then
             end_s = "custom"
+        else
+            when( end_s == "truncate"
+                eext = 0
+            )
         )
```
Comment (sporrasm) [inline run_scripts/start_bag.il]: I'm not sure about the origin or reason behind these changes and thus they were removed in the next commit (04322db).

**Comment by ayan-biswas** [comment]
> @ayan-biswas Just to check, is master your current development branch or should I select some other target?

Yes, master is the latest branch, and we don't see these issues. We just have to refresh Virtuoso sometimes to see all the updates.

**Comment by mkosunen** [comment]
> Yes, master is the latest branch, and we don't see these issues. We just have to refresh Virtuoso sometimes to see all the updates.

@ayan-biswas I see, these features are only visible if you use parametrized pcells (like vendor provided transistor device generators or resistor generators)   in a hierarchical design: parametrized pcell inside a templateDB inside a templateDB, AND the pcell requires CDF callbacks to be run. I do not know how much you have this kind of structures as for example when using analogBase  you usually have  none of these. 

Anyway, as tested, the proposed fix works with the hierarchical design too, and otherwise the functionality is the same as before. 

(we fare not using cybagoa, and I do not know if that affects something.)

**Comment by mkosunen** [comment]
@sporrasm Just to avoid unnecessary modifications/branch divergence, can you check if just refreshign Virtuoso without this hack would bring us the desired operation with one of our desings, and re-confirm whether we need this fix or not?

**Comment by sporrasm** [comment]
> @sporrasm Just to avoid unnecessary modifications/branch divergence, can you check if just refreshign Virtuoso without this hack would bring us the desired operation with one of our desings, and re-confirm whether we need this fix or not?

Just tested it, unfortunately it doesn't seem to do the trick for us for some reason..

File: run_scripts/start_bag.il [hunk 4]
```diff
@@ -1441,15 +1458,6 @@ procedure( create_layout( lib_name view_name via_tech layout_f "ttt" )
             printf("Creating %s__%s (%s)\n" lib_name cell_name view_name)
             create_layout_helper(cv tech_file inst_list rect_list via_list pin_list path_list
                                  blockage_list boundary_list polygon_list)
-            
-            if( exists( inst inst_list inst->params != nil ) ; Check whether cellview contains any Pcell instances
-                if( getd('CCSinvokeCdfCallbacks) != nil then ; Found function handle for callbacks, run callbacks 
-                    printf("Running callbacks for cellview: %s\n" cell_name)
-                    (CCSinvokeCdfCallbacks cv)
```
Comment (sporrasm) [inline run_scripts/start_bag.il]: This way of invoking callbacks seems to update layouts perfectly fine. However, it takes an excessive amount of time to create layouts.

File: run_scripts/start_bag.il [hunk 5]
```diff
@@ -1256,6 +1265,14 @@ procedure( create_layout_helper( cv tech_file inst_list rect_list via_list pin_l
                         obj = dbCreateParamSimpleMosaic(cv inst_cv inst->name inst->loc inst->orient
                                                         inst->num_rows inst->num_cols inst->sp_rows inst->sp_cols
                                                         inst->params)
+                        when( obj
+                            if( inst->param_order != nil then
+                                param_order=inst->param_order
+                            else
+                                param_order= mapcar( lambda( (x) car(x) ) inst->params)
+                            )
+                            CCSinvokeInstCdfCallbacks(obj ?order param_order)
```
Comment (sporrasm) [inline run_scripts/start_bag.il:1274]: This method of calling the callbacks seems to be a lot faster. The only difference from the master branch seems to be that the method used for callbacks is changed from CCSinvokeCdfCallbacks() to CCSinvokeInstCdfCallbacks(). This seems to work on our side (e.g. layouts get updated correctly AND generation speed is much faster).

**Pull request merged**